lib/support/reconstructPath

```diff
@@ -0,0 +1,23 @@
+/**
+ * Reconstructs the path for the current key, from currentPath (requestedPath)
+ * state passed maintained during walk operations.
+ *
+ * During the walk, since the requestedPath array, and requestedPath.index
+ * is updated after we attempt to merge/insert nodes during a walk (it reflects
+ * the inserted node's parent branch) we need to reconstitute a path from it.
+ *
+ * @param  {Array} currentPath The current requestedPath state, during the walk
+ * @param  {String} key        The current key value, during the walk
+ * @return {Array} A new array, with the path which represents the node we're about
+ * to insert
+ */
+module.exports = function reconstructPath(currentPath, key) {
+
+    // The first time called during a walk, currentPath.index won't be set.
+    var currentPathIndex = (currentPath.index || 0) + 1;
```

Comment:
I agree. I'm happy to feed in index, but it seems like those are heavily coupled already (requestedPath and requestedPath.index always have to by in sync don't they?), so separating them here seems riskier, unless we separate them all the way up.